### ISSUE TITLE
How will EnglishScript be translated into other programming languages?

### ISSUE THREAD

**Issue by jarble** [pannous/english-script#2]
One of the long-term goals of EnglishScript is to make the language available on multiple platforms. I have written a program called [transpiler-generator](https://github.com/jarble/transpiler-generator) that is intended to translate EnglishScript into several other programming languages, and vice-versa.

Which features will I need to implement in order to make transpiler-generator compatible with EnglishScript?

**Comment by jarble** [comment]
Also, transpiler-generator will be able to convert EnglishScript into C and C++ if the EnglishScript source code is statically typed. Is it possible yet to write statically-typed code in EnglishScript?

**Comment by pannous** [comment]
Nice!

EnglishScript / [Angle](https://github.com/pannous/angle) parses into a [canonical universal AST](https://github.com/pannous/kast) which is meant to be language agnostic. From that Abstract syntax tree we generate bytecode. The [kast](https://github.com/pannous/kast) subproject is work in progress, but it was already used to transpile the Ruby version to the better Python version.

It would be very sweet if we had a kast -> [transpiler](https://github.com/jarble/transpiler-generator) adapter!

**Comment by pannous** [comment]
>>  Is it possible yet to write statically-typed code in EnglishScript?
Absolutely!

 types are optional but if they are used they will be enforced:
x=8
x="hi" // ok
int x=9
x="hi" // error, x typed as int

same for function arguments

[tests](https://github.com/pannous/angle/blob/master/tests/type_test.py)

**Comment by pannous** [comment]
For a discussion on automatic global static typing also see https://github.com/manastech/crystal/issues/1824

**Comment by pannous** [comment]
Here was the first draft for a [schema](https://github.com/pannous/kast/blob/master/kast.xsd), however out of lazyness we sticked to the python AST for now. It would be nice though if we could map 
```XML
    <complexType name="method">
        <choice maxOccurs="unbounded">
            <element name="name" type="string"/>
            <element name="returns" type="a:type"/>
            <element name="modifiers" type="a:modifiers"/>
            <element name="arguments" type="a:arguments"/>
            <element name="body" type="a:block"/>
            <!-- IMPLICIT: BODY!!-->
        </choice>
```
to your
`anonymous_function,params:function_parameters,b:series_of_statements,return_type:type`

**Comment by jarble** [comment]
I'm also planning to write a translator that converts Java source code into statically typed EnglishScript code. Has the syntax for statically typed parameters been decided yet?

**Comment by jarble** [comment]
I'm also planning to write a translator that converts Java and C code into statically typed EnglishScript source code. Has the syntax for statically typed parameters been decided yet?

**Comment by pannous** [comment]
Yes, EnglishScript parses, Java, C#, python and ruby style methods, as well as it's [own syntax](
https://github.com/pannous/english-script-samples/blob/master/fibonaccis.e) and probably some [verbose variants](https://github.com/pannous/english-script-samples/blob/master/fibonaccis2.e)

**Comment by pannous** [comment]
here is the grammar for the method definition:
```python

@Starttokens(method_tokens)
def method_definition(name=None):
  # annotations=maybe(annotations)
  # modifiers=maybe(modifiers)
  if not name:
    tokens(method_tokens)  # how to
    no_rollback()
    name = word(include=english_operators)  # maybe(noun) or verb()  # integrate or word
  # obj= maybe( endNode ) # a sine wave  TODO: invariantly get as argument book.close==close(book)
  brace = maybe_token('(')
  args = []

  def arguments():
    angle.in_params = True
    a = param(len(args))
    maybe_token(',')
    args.append(a)
    return a

  star(arguments)  # i.e. 'over an interval i' 'from a to b' 'int x, int y=7'

  return_type = maybe_tokens(['as']) and maybe(typeNameMapped) or None
  return_type = maybe_tokens(['return', 'returns', 'returning', '=', '->']) and maybe(typeNameMapped) or return_type
```

**Comment by pannous** [comment]
generous [method_tokens](https://github.com/pannous/angle/blob/master/core/english_tokens.py#L27)

**Comment by pannous** [comment]
once you get to generics we need to discuss again.
Vector<Integer> should probably be mapped to EnglishScript "list of numbers"

**Comment by pannous** [comment]
the preferred **verbose** typed method declaration would be
`define fibonacci number n as integer:
  <block>
`
"as integer" specifying the return type

**Comment by pannous** [comment]
it's kind of fascinating that the same grammar and bath parses these stylistically quite different complete method definitions:
`define addition of number a and number b as number return a + b`
`to add a to b: a + b`
